### PR TITLE
Add stress tests for settings table and fix simple issues with transactions

### LINKED ISSUE #1282: Add stress tests for settings table
* [x] Add a stress integration test that creates X collections in parallel in X goroutines.
* [ ] As the test will always fail and implementation is not in scope, either skip the test or not merge it.

### PR DESCRIPTION
**Pull request by rumyantseva**
# Description

Closes #1282. 

See also my comments in the original issue.
See also #1318 for `SchemaStats` related issues.

## Readiness checklist

* [x] I added tests for new functionality or bugfixes.
* [ ] I ran `task all`, and it passed.
* [ ] I added/updated comments for both exported and unexported top-level declarations (functions, types, etc).
* [ ] I checked comments rendering with `task godocs`.
* [x] I ensured that the title is good enough for the changelog.
* [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Assignee, Labels, Project and project's Sprint fields.
* [x] I marked all done items in this checklist.

### CODE REVIEW

File: integration/create_test.go [hunk 1]
```diff
@@ -26,6 +27,43 @@ import (
...
+	wg.Wait()
+
+	// check that all collections were created, and we can query them
+	for i := 0; i < collNum; i++ {
+		t.Run(fmt.Sprintf("check_stress_%d", i), func(t *testing.T) {
+			t.Parallel()
+
+			collName := fmt.Sprintf("stress_%d", i)
+			res := db.Collection(collName).FindOne(ctx, bson.D{})
+			require.Equal(t, mongo.ErrNoDocuments, res.Err())
```
Comment (chilagrow) [inline integration/create_test.go:62]: For a non-existing collection, we do not get `ErrNoDocuments`, correct? Just want to ensure that `ErrNoDocuments` is an error we get if the collection exists and no document in the collection.

File: internal/handlers/pg/msg_create.go [hunk 2]
```diff
@@ -72,25 +72,44 @@ func (h *Handler) MsgCreate(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg,
...
+	// but it's a normal situation, and we should create both collections in such a case.
+
 	err = h.PgPool.InTransaction(ctx, func(tx pgx.Tx) error {
 		if err := pgdb.CreateDatabaseIfNotExists(ctx, tx, db); err != nil {
-			if errors.Is(pgdb.ErrInvalidDatabaseName, err) {
+			switch {
+			case errors.Is(err, pgdb.ErrAlreadyExist):
+				// If the DB was created from a parallel query, it's ok.
+				// However, in this case one of the transaction queries failed,
+				// so we need to rollback the transaction.
```
Comment (chilagrow) [inline internal/handlers/pg/msg_create.go:86]: Nice catch!

File: integration/create_test.go (hunk 1 above)
Comment (rumyantseva) [reply integration/create_test.go:62]: Good point! If collection doesn't exist, it gives `ErrNoDocuments` 🙈

**Comment by rumyantseva** [comment]
Ooops! Not ready for review yet!

File: integration/create_test.go (hunk 1 above)
Comment (rumyantseva) [reply integration/create_test.go:62]: That actually revealed some other problems, see the recent update :)

File: integration/create_test.go [hunk 3]
```diff
@@ -26,6 +27,57 @@ import (
 	"github.com/FerretDB/FerretDB/integration/setup"
 )
 
+func TestCreateStressPostgres(t *testing.T) {
+	setup.SkipForPostgresWithReason(t, "https://github.com/FerretDB/FerretDB/issues/1206")
+	setup.SkipForTigrisWithReason(t, "Tigris needs a schema for collection creation")
```
Comment (AlekSi) [inline integration/create_test.go:32]: Could we add a schema and make this test handler-independent?

File: integration/create_test.go [hunk 4]
```diff
@@ -26,6 +27,57 @@ import (
...
+		wg.Add(1)
+
+		go func(i int) {
+			collName := fmt.Sprintf("stress_%d", i)
+			err := db.CreateCollection(ctx, collName)
+			_, errIns := db.Collection(collName).InsertOne(ctx, bson.D{{"_id", "foo"}, {"v", "bar"}})
+
+			wg.Done()
+
+			require.NoError(t, err)
```
Comment (AlekSi) [inline integration/create_test.go:52]:
```suggestion
			assert.NoError(t, err)
```

`require` does not work correctly from non-main goroutine: https://pkg.go.dev/testing#T.FailNow

File: integration/create_test.go [hunk 5]
```diff
@@ -26,6 +27,57 @@ import (
...
+	var wg sync.WaitGroup
+	for i := 0; i < collNum; i++ {
+		wg.Add(1)
+
+		go func(i int) {
+			collName := fmt.Sprintf("stress_%d", i)
+			err := db.CreateCollection(ctx, collName)
+			_, errIns := db.Collection(collName).InsertOne(ctx, bson.D{{"_id", "foo"}, {"v", "bar"}})
+
+			wg.Done()
```
Comment (AlekSi) [inline integration/create_test.go:50]:
```suggestion
			defer wg.Done()
```

And move on top

File: integration/create_test.go [hunk 6]
```diff
@@ -26,6 +27,57 @@ import (
...
+func TestCreateStressPostgres(t *testing.T) {
+	setup.SkipForPostgresWithReason(t, "https://github.com/FerretDB/FerretDB/issues/1206")
+	setup.SkipForTigrisWithReason(t, "Tigris needs a schema for collection creation")
+
+	t.Parallel()
+
+	ctx, collection := setup.Setup(t) // no providers there, we will create collections concurrently
+	db := collection.Database()
+
+	collNum := 10
```
Comment (AlekSi) [inline integration/create_test.go:39]:
```suggestion
	const collNum = runtime.GOMAXPROCS() * 10
```

File: integration/create_test.go [hunk 7]
```diff
@@ -26,6 +27,57 @@ import (
...
+
+	var wg sync.WaitGroup
+	for i := 0; i < collNum; i++ {
+		wg.Add(1)
+
+		go func(i int) {
+			collName := fmt.Sprintf("stress_%d", i)
```
Comment (AlekSi) [inline integration/create_test.go:46]: I think we need an extra trigger to start (not just create) all goroutines at the same time. Something like:

```go
ready := make(chan struct{}, collNum)
start := make(chan struct{})

// ...
go func(i int) {
  ready <- struct{}
  <-start
  // ...
}

// ...

for i := 0; i < collNum; i++ {
  <- ready
}

close(start)
```

File: integration/create_test.go [hunk 8]
```diff
@@ -26,6 +27,57 @@ import (
...
+		}(i)
+	}
+
+	wg.Wait()
+
+	colls, err := db.ListCollectionNames(ctx, bson.D{})
+	require.NoError(t, err)
+
+	// TODO https://github.com/FerretDB/FerretDB/issues/1206
+	// This check fails now. Even though all the collections are created as separate tables in the database,
```
Comment (AlekSi) [inline integration/create_test.go:63]: Well, it doesn't, because we skip that test for both PostgreSQL and Tigris

File: integration/create_test.go [hunk 9]
```diff
@@ -26,6 +27,57 @@ import (
...
+	// the settings table doesn't store all of them because of concurrency issues.
+	require.Len(t, colls, collNum)
+
+	// check that all collections were created, and we can query them
+	for i := 0; i < collNum; i++ {
+		t.Run(fmt.Sprintf("check_stress_%d", i), func(t *testing.T) {
```
Comment (AlekSi) [inline integration/create_test.go:69]: `i := i` to make an iteration-specific copy of `i`

File: integration/create_test.go [hunk 10]
```diff
@@ -26,6 +27,57 @@ import (
...
+			t.Parallel()
+			collName := fmt.Sprintf("stress_%d", i)
```
Comment (AlekSi) [inline integration/create_test.go:71]: otherwise, that `i` could be incorrect in parallel subtest

**Review by AlekSi — COMMENTED** [review]
tbh, I'm quite confused how things are split between this PR and #1318: both add tests, tweak tests, skip tests, and fix code. Let's make one PR with this stress test that is not skipped and fixes that make it pass

File: integration/create_test.go (hunk 7 above)
Comment (rumyantseva) [reply integration/create_test.go:46]: I understand the idea with `start`, but what is `ready` for? Is it to replace `wg`?

Comment (AlekSi) [reply integration/create_test.go:46]: To start all created goroutines when they are ready, not just when they are created.

We could use a wait group for that, but a different variable.

File: integration/create_test.go (hunk 9 above)
Comment (rumyantseva) [reply integration/create_test.go:107]: fixed

File: integration/create_test.go (hunk 10 above)
Comment (rumyantseva) [reply integration/create_test.go:110]: fixed

File: integration/create_test.go (hunk 8 above)
Comment (rumyantseva) [reply integration/create_test.go]: fixed

File: integration/create_test.go (hunk 4 above)
Comment (rumyantseva) [reply integration/create_test.go]: fixed

File: integration/create_test.go (hunk 6 above)
Comment (rumyantseva) [reply integration/create_test.go]: This is unclear to me, could you explain the idea behind it?

**Review by chilagrow — DISMISSED** [review]
Looks good!

File: integration/create_test.go [hunk 11]
```diff
@@ -26,6 +29,94 @@ import (
...
 
+func TestCreateStress(t *testing.T) {
+	setup.SkipForPostgresWithReason(t, "https://github.com/FerretDB/FerretDB/issues/1206")
+
+	t.Parallel()
+
+	ctx, collection := setup.Setup(t) // no providers there, we will create collections concurrently
+	db := collection.Database()
+
+	collNum := runtime.NumCPU() * 10
```
Comment (AlekSi) [inline integration/create_test.go]: That should be `runtime.GOMAXPROCS(-1)` as we are interested in the number of CPU cores ("procs") that the Go scheduler is allowed to use, not just in the number of CPU cores

File: integration/create_test.go [hunk 12]
```diff
@@ -26,6 +29,94 @@ import (
...
+			}`, collName, i,
+			)
+			opts := options.CreateCollectionOptions{
+				Validator: bson.D{{"$tigrisSchemaString", schema}},
+			}
+
+			// Attempt to create a collection for Tigris with a schema.
+			// If we get an error, it's not Tigris, so we create collection without schema
+			err := db.CreateCollection(ctx, collName, &opts)
+			if err != nil {
```
Comment (AlekSi) [inline integration/create_test.go:75]: Well, that could also be a concurrency issue that we just ignore, defeating the purpose of the test.

I think we should check that error. We already do that in one other test

File: internal/handlers/tigris/tigrisdb/databases.go [hunk 13]
```diff
@@ -37,19 +39,32 @@ func (tdb *TigrisDB) createDatabaseIfNotExists(ctx context.Context, db string) (
...
 	// but keep in mind that it can be created in concurrent connection.
-	err = tdb.Driver.CreateDatabase(ctx, db)
-	switch err := err.(type) {
-	case nil:
-		return true, nil
-	case *driver.Error:
-		if IsAlreadyExists(err) {
-			return false, nil
-		}
+	// If we detect that other creation is in flight, we give up to three attempts to create the database.
```
Comment (AlekSi) [inline internal/handlers/tigris/tigrisdb/databases.go:42]: Do we want to keep it, or we want Tigris to fix it?

File: internal/handlers/tigris/tigrisdb/errors.go [hunk 14]
```diff
@@ -34,6 +37,22 @@ func IsAlreadyExists(err error) bool {
 	return pointer.Get(e).Code == api.Code_ALREADY_EXISTS //nolint:nosnakecase // Tigris named their const that way
 }
 
+// IsOtherCreationInFlight returns true if an attempt to create the database with the given name is already in progress.
+// This function is implemented to keep nolint in a single place.
+func IsOtherCreationInFlight(err error) bool {
```
Comment (AlekSi) [inline internal/handlers/tigris/tigrisdb/errors.go:42]: Let's un-export it

File: integration/create_test.go (hunk 11 above)
Comment (AlekSi) [reply integration/create_test.go]: Doh, too late!

File: internal/handlers/tigris/tigrisdb/databases.go (hunk 13 above)
Comment (rumyantseva) [reply internal/handlers/tigris/tigrisdb/databases.go:42]: I'm afraid we have to keep it. Tigris said they would fix the error code (so it's not "unknown" anymore). But as far as I understand, they consider such a situation possible - the DB is in the creation process from some other request, and we don't know yet if it'd be created successfully. So we need another attempt to understand how it went.

File: integration/create_test.go (hunk 12 above)
Comment (rumyantseva) [reply integration/create_test.go:77]: fixed

File: internal/handlers/tigris/tigrisdb/errors.go (hunk 14 above)
Comment (rumyantseva) [reply internal/handlers/tigris/tigrisdb/errors.go]: fixed

File: internal/handlers/tigris/tigrisdb/databases.go (hunk 13 above)
Comment (AlekSi) [reply internal/handlers/tigris/tigrisdb/databases.go:42]: My understanding is that their client would do retrying transparently for us.

Let's create an issue for that in our repo and add TODO there

Comment (rumyantseva) [reply internal/handlers/tigris/tigrisdb/databases.go:42]: added

File: integration/create_test.go [hunk 15]
```diff
@@ -26,6 +31,100 @@ import (
...
+					"v": {"type": "string"}
+				}
+			}`, collName, i,
+			)
+			opts := options.CreateCollectionOptions{
+				Validator: bson.D{{"$tigrisSchemaString", schema}},
+			}
+
+			// Attempt to create a collection for Tigris with a schema.
+			// If we get an error, it's not Tigris, so we create collection without schema
```
Comment (AlekSi) [inline integration/create_test.go:75]: > it's not Tigris

Well, it is never Tigris. It is either FerretDB or MongoDB. And I guess only MongoDB returns that error?

**Review by AlekSi — DISMISSED** [review]

File: integration/create_test.go (hunk 15 above)
Comment (rumyantseva) [reply integration/create_test.go:75]: @AlekSi, that's true, but I'm not sure how to write such a comment, could you rephrase it, so it looks better?

Comment (AlekSi) [reply integration/create_test.go:75]: > If we get an error, that's MongoDB (FerretDB ignores that argument for non-Tigris handlers), so we create collection without schema

**Review by w84thesun — APPROVED** [review]
LGTM

**Review by chilagrow — APPROVED** [review]

**Review by noisersup — APPROVED** [review]

File: integration/create_test.go [hunk 16]
```diff
@@ -26,6 +31,101 @@ import (
...
+			}
+
+			// Attempt to create a collection for Tigris with a schema.
+			// If we get an error, that's MongoDB (FerretDB ignores that argument for non-Tigris handlers),
+			// so we create collection without schema.
+			err := db.CreateCollection(ctx, collName, &opts)
+			if err != nil {
+				var cmdErr *mongo.CommandError
+				if errors.As(err, &cmdErr) {
+					if strings.Contains(cmdErr.Message, `unknown top level operator: $tigrisSchemaString`) {
```
Comment (AlekSi) [inline integration/create_test.go:81]: We have the same comment there and at https://github.com/FerretDB/FerretDB/pull/1318#discussion_r1009398501, but the implementation of error checking is very different. Which one is correct?
